clang/lib/Sema/SemaBounds.cpp

```diff
@@ -1827,6 +1828,16 @@ namespace {
       // Is R a subrange of this range?
       ProofResult InRange(BaseRange &R, ProofFailure &Cause, EquivExprSets *EquivExprs,
                           std::pair<ComparisonSet, ComparisonSet>& Facts) {
+
```

Comment:
We talked about this in person.   We agreed that we want to allow empty bounds declarations: they might represent a zero-element array or an empty buffer.   We want to disallow memory access to empty ranges.  We may want to statically disallow ranges that are invalid (where the upper bound < lower bound).    There are some algorithms like quicksort where pointers intentionally cross - I'm wondering if that leads to "invalid ranges" in some formulations of quicksort.

Given this, I think it makes sense to make sure there is a check for an empty range in CheckBoundsAtMemoryAccess.  It doesn't matter what the actually access expression is.   It would also make sense at bounds declarations, calls, and static casts to check for invalid ranges.    I believe that should be done as a separate check on destination bounds for calls and bounds casts.   For variables, I believe you can add the check to VisitVarDecl (declared around line 3228).    At these locations, it will also be easiere to distinguish between ounds for nt_array_ptr and array_ptr.

Reply:
Previously we were disallowing LB == UB cases, which was problematic for nt_arrays. But now that we are allowing this case, is there any difference in the handling of nt_array_ptr vs. array_ptr?